Change the place of DLL resolve hook

ryan-linehan opened the pull request:
Seemingly fixes #5. Unsure of all the impacts it might have? But, godot was not loading steamworks correctly without this when using AnyCPU

ryan-linehan committed:
include STEAMWORKS_WIN constant

ryan-linehan commented:
It seems that with this change it breaks the mac/linux builds. As I tried the fix on my arm mac and it ends up looking for _64 dlls that don't exist. I reverted this change and deployed again to my arm mac and it fixed it. Investigating further before reopening. This also might be what #1 is trying to fix?

ryan-linehan committed:
add a fix to dynamically load assembly on windows

Akarinnnnn commented:
Defining both `STEAMWORKS_WIN` and `STEAMWORKS_LIN_OSX` can potentially break some internal used values. I suggest don't do it.

In case of `NativeMethods.cs`, this change looks like a good approach that we can apply hook for all `AssemblyLoadContext`s. Try to reference `static class AnyCPU` in `Steam.cs` for writing hook. Once your hook is prove to work, I will help you to clean legacy hook up.

By the way, you may not noticed that `NativeMethods.cs` is completely **auto generated** with a template. Once you prove your hook is working, I can help you to moving it to template file.

Akarinnnnn committed:
Revert adding of macro `STEAMWORKS_WIN`

To avoid breaking internal symbol usages, `STEAMWORKS_WIN` is dropped versus `STEAMWORKS_LIN_OSX`.

Akarinnnnn committed:
Changed the time to hook dll loading.

Now we can do our hook in any AssemblyLoacContext s. DLL loading issues in godot(editor?)  caused by multiple A.L.C.s should be fixed.

ryan-linehan commented:
> By the way, you may not noticed that `NativeMethods.cs` is completely **auto generated** with a template. Once you prove your hook is working, I can help you to moving it to template file.

Looks like it is mostly working so I think we should go ahead and do that. I was able to get steamworks to initialize on both windows and my arm64 mac. So I believe you could do achievements and the such through the .AnyCPU package with how this is currently working after this change. I was also able to get players to join a lobby together in my game using Steamworks.NET. 

From mac:

<img width="1166" height="1152" alt="image" src="https://github.com/user-attachments/assets/7cfbd9a7-6d49-491f-a38e-46aa6762e8ee" />

From windows:
<img width="1158" height="697" alt="image" src="https://github.com/user-attachments/assets/baaa90f6-d88b-41c3-bda8-137304d5e970" />

> note: the peer status in the lobby screen is basically a single player peer status. No one has attempted to establish a connection with each other anywhere but steam lobby requests. Until the host hits 'start' game it is all just lobby requests through steam

However, when my players attempt to connect to each other through steams p2p networking that I have setup. You can see an error printed out in the mac image. I am unsure what or why this is. Once I try to get the second client to try and establish a connection to the host it just stays in connecting state:
 
<img width="1152" height="690" alt="image" src="https://github.com/user-attachments/assets/b54025f7-c341-453e-a863-0927e0222e62" />

I have successfully used my nuget package created in this [repo](https://github.com/[ryan-linehan/Godot.Steamworks.NET](https://github.com/ryan-linehan/Godot.Steamworks.NET)) to connect through steams p2p networking with the same game code on windows to windows machines. But, after testing on two windows machines with a nuget package from .AnyCPU the clients get stuck in a pending state (see above). Any ideas on why this might be? 

Let me know if I can help. My [repository]((https://github.com/[ryan-linehan/Godot.Steamworks.NET]) might be useful in quickly being able to test the P2P setup for a steam game. Happy to help however I can

Thanks!

Akarinnnnn committed:
Move loader hook fix to source generator's template.
